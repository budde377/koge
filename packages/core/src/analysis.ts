
export type BaseContext = Record<string, unknown>

export type BaseAttributes = Record<string, unknown>

export type Analysis<TContext = BaseContext, TAttributes = BaseAttributes, TInput = unknown> = (input: TInput, rootContext: TContext) => TAttributes

export type AnalysisContextBuilder<TAstNode, TContext, TAttributes> = (node: TAstNode, c: () => TContext, attr: (node: TAstNode) => TAttributes) => TContext

export type AnalysisAttributeBuilder<TAstNode, TContext, TAttributes> =  (node: TAstNode, c: () => TContext, attr: (node: TAstNode) => TAttributes) => TAttributes

export function analysisBuilder<TAstNode, TContext extends BaseContext = BaseContext, TAttributes extends BaseAttributes = BaseAttributes, TInput = TAstNode> (
  findChildren: (n: TAstNode) => TAstNode[], 
  parse: (f: TInput) => TAstNode,
  attributeBuilder: AnalysisAttributeBuilder<TAstNode, TContext, TAttributes>,
  contextBuilder?: AnalysisContextBuilder<TAstNode, TContext, TAttributes>,  
): Analysis<TContext, TAttributes, TInput> {
  return (input, rootContext) => {
    const node = parse(input)
    const contexts:  Map<TAstNode, TContext> = new Map()
    const attributes: Map<TAstNode, TAttributes> = new Map()
    const parents: Map<TAstNode, TAstNode> = new Map()
    function attrLookup(n: TAstNode): TAttributes {
      const attrs = attributes.get(n)
      if (attrs != null) return attrs
      const newAttrs = attributeBuilder(n, ctxLookup(n), attrLookup)
      attributes.set(n, newAttrs)
      return newAttrs
    }

    function ctxLookup(n: TAstNode): () => TContext {
      if (!contextBuilder) return () => rootContext
      return () => {
        const ctx = contexts.get(n)
        if (ctx != null) return ctx
        const parentN = parents.get(n)
        const newCtx = parentN ? contextBuilder(n, ctxLookup(parentN), attrLookup) : rootContext
        contexts.set(n, newCtx)
        return newCtx
      }
    }

    const stack = []
    let n: TAstNode | undefined = node
    while (n) {
      const chldrn = findChildren(n)
      for(let i = chldrn.length - 1; i >= 0; i --) {
        const child = chldrn[i]
        stack.push(child)
        parents.set(child, n)
      }
      n = stack.pop()
    }
    return attributeBuilder(node, ctxLookup(node), attrLookup)
  }
}