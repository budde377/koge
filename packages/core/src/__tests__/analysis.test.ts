import { AnalysisAttributeBuilder, analysisBuilder, AnalysisContextBuilder, BaseAttributes, BaseContext } from "../analysis"

type Node = {kind: 'node', children: Ast[]}
type Leaf =  {kind: 'leaf', value: string}

type Ast =
| Node
| Leaf

function analysis<TContext extends BaseContext, TAttributes extends BaseAttributes>(
  attributeBuilder: AnalysisAttributeBuilder<Ast, TContext, TAttributes>,
  contextBuilder?: AnalysisContextBuilder<Ast, TContext, TAttributes>
) {
  return analysisBuilder<Ast, TContext, TAttributes>(
    (n) => n.kind === 'node' ? n.children : [],
    (n) => n,
    attributeBuilder,
    contextBuilder
  )
}

function node(children: Ast[] = []): Node {
  return {kind: 'node', children}
}

function leaf(value: string): Leaf {
  return {kind: 'leaf', value}
}

describe('analysis', () => {
  const printAnalysis = analysis<{indent: string}, {result(): string}>(
    (n, context, attr) => ({
      result(): string {
        const c = context()
        if(n.kind === 'leaf'){
          return `${c.indent}${n.value}\n`
        } else {
          return n.children.map((c: Ast) => attr(c).result()).join('')
        }
      }
    }),
  (n, c) => {
    const ctx = c()
    if (n.kind === 'node') return {indent: ctx.indent + "  "}
    return ctx
  }
  )

  it('will pass context', () => {
    expect(printAnalysis({
      kind: 'leaf'  ,
      value: 'foobar',
    }, {indent: ''}).result()).toMatchSnapshot()
  })

  it('will print complex', () => {
    expect(printAnalysis(
      node([
        node([
          leaf('Hello'),
          node([
            leaf('World'),
            node(
              [
                leaf('how'),
                leaf('is'),    
              ]
            ),
            leaf('life!')
          ]),
          node()
        ])
      ])
      , {indent: ''}).result()).toMatchSnapshot()
  })
})