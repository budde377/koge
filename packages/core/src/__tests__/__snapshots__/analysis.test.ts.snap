// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`analysis will pass context 1`] = `
"foobar
"
`;

exports[`analysis will print complex 1`] = `
"  Hello
    World
      how
      is
    life!
"
`;
